Sprint7 task3 import rewards csv

iAcay opened the pull request:
Hey! The code which is below was implemented in order to meet the acceptance criterias of the third task in sprint no. 7 :)

AC in this task:
-- On a dedicated view there is a form that allows the admin to upload a CSV with the reward title, description, price and category
-- Reward's title should be unique
-- When a reward exists its description, price and category should be updated
-- When an error happens an admin should see why import failed
-- There is a system spec testing uploading a file

In order to speed up bulk inserting data I use Activerecord-Import gem, which is supposedly much faster than standard ActiveRecord save. That gem is quite popular (60mln+ downloads). Link to its documentation --> https://github.com/zdennis/activerecord-import

https://user-images.githubusercontent.com/96024046/183976145-aabe2e92-85ad-4ef9-b54d-bd24c173fc31.mov

iAcay committed:
feat: possibility to display multiline alerts

iAcay committed:
feat: unique constraint for rewards' titles

iAcay committed:
feat: add 'activerecord-import' gem

iAcay committed:
feat: possibility to import rewards from a csv file

-- add 'new_import_from_csv' get route and 'import_from_csv' post route,
-- implement new_import_from_csv and import_from_csv methods to admin/rewards_controller
-- add dedicated form for uploading CSV files with rewards' data
-- add link to above mentioned form into admin_rewards_index page

iAcay committed:
tests: can admin import rewards from csv

-- checks if admin can import reward from csv file
-- checks if reward's description, price and category are updated when imported csv file includes already existing reward
-- checks validations

maikhel commented on db/migrate/20220810113531_add_unique_to_rewards_title.rb:
```diff
@@ -0,0 +1,5 @@
+class AddUniqueToRewardsTitle < ActiveRecord::Migration[6.1]
```
I think `AddUniqueIndexOnTitleToRewards` would be more informative

iAcay replied:
First of all, thanks a lot for review! I have changed this according to your suggestion :)

maikhel commented on spec/files/test1.csv:
```diff
@@ -0,0 +1,2 @@
+Title,Description,Price,Category
+RewardTitle,RewardDescription,3,Category1
```
Would be good to have two or three rows to import

iAcay replied:
Good point! I have added additional rows 👍

maikhel commented on spec/files/test2.csv:
```diff
@@ -0,0 +1,2 @@
+Title,Description,Price,Category
```
Usually there is a directory: `spec/fixtures` with such files. Then you could have more specific directory for rewards and more informative names for files, eg:
```
spec/fixtures/rewards/rewards_import.csv
spec/fixtures/rewards/invalid_rewards_import.csv

```

iAcay replied:
Understood! I have created folder fixtures and moved testing files there. I have also changed .csv files' names to more informative ;)

maikhel commented on spec/system/admin_rewards_from_csv_spec.rb:
```diff
@@ -0,0 +1,65 @@
+      attach_file(File.absolute_path('./spec/files/test1.csv'))
+      click_button 'Import Rewards'
+
+      expect(page).to have_content 'Rewards was successfully imported.'
```
typo, `Rewards were...`

iAcay replied:
My bad... Corrected 🖌️

maikhel commented on app/controllers/AdminUsers/rewards_controller.rb:
```diff
@@ -38,6 +40,37 @@ def destroy
+        redirect_to new_import_from_csv_admin_users_rewards_path, alert: 'Only .csv files are allowed.'
+      else
+        begin
+          ActiveRecord::Base.transaction do
```
This could be moved to separate class (eg. service for importing).

iAcay replied:
Done 👍 I have implemented seperate service for importing :)

maikhel commented on app/controllers/AdminUsers/rewards_controller.rb:
```diff
@@ -38,6 +40,37 @@ def destroy
+              category = Category.find_or_create_by(title: row[3].capitalize)
+              values << [row[0], row[1], row[2], category.id]
+            end
+            Reward.import! columns, values, validate: true,
```
TBH, I'm not sure if using `activerecord-import` gem helps here - you still have to find category and people not familiar with it would need to read how it specifically works (eg. what `conflict_target` option does). Without `activerecord-import` you could have:
```
reward = Reward.find_or_initialize_by(title: row['Title'])
reward.description = row['Description']
reward.price = row['Price']
reward.category = Category.find_or_create_by!(title: row['Category'])
reward.save!
```
Usually, it's better to use build-in functions and not add third-party dependencies unless really necessary.

iAcay replied:
Understood! I have already removed `activerecord-import` gem and put in this place the code suggested by you. Thanks! 👍

maikhel commented on spec/system/admin_rewards_from_csv_spec.rb:
```diff
@@ -0,0 +1,65 @@
+      click_link 'Manage Rewards'
+
+      click_link 'Import rewards from CSV'
+      attach_file(File.absolute_path('./spec/files/test1.csv'))
```
For such imports, try to import more than one row/record to be sure loop works correctly.

iAcay replied:
Done!

maikhel commented on spec/system/admin_rewards_from_csv_spec.rb:
```diff
@@ -0,0 +1,65 @@
+      expect(page).to have_content '3'
+      expect(page).to have_content 'Category1'
+
+      # WHEN REWARD WITH IMPORTED TITLE EXISTS ITS DESCRIPTION, PRICE AND CATEGORY SHOULD BE UPDATED
```
It could be a new test, like `it 'updates existing rewards during import'`. Then you don't need this comment :)

iAcay replied:
Done! 👍

maikhel reviewed (approved):
Looks good!

Btw, try to use "imperative mood" in your commit messages - they will be easier to read after some time. More tips:
https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/

iAcay committed:
style: typo correction in notices

iAcay committed:
refactor: change file name from AddUniqueToRewardsTitle to  AddUniqueIndexOnTitleToRewards

iAcay committed:
tests: add more rows in csv files for testing

- move .csv files from spec / files folder into spec / fixtures / rewards folder
- improve .csv files names
- add additional rows to .csv files for importing rewards
- adjust tests in relation to adding lines to csv files

iAcay committed:
refactor: remove 'activerecord-import' gem

iAcay committed:
refactor: move the code responsible for importing rewards to a separate service object
